site/docs/references/custom-selectors.md

```diff
@@ -0,0 +1,164 @@
+
+```css
+@namespace "Page";
+:import {
+    -st-from: "./comp.st.css";
+    -st-default: Comp;
+}
+/*
+selector: .Page__root .Comp__root .Comp__controls .Comp__btn
+*/
+Comp::controlBtn { 
+    background: gold; 
+}
+```
+
+> **Notice**:  
+> In case a `custom-selector` alias conflicts with a local CSS class name, the exposed `pseudo-element` will target the `custom-selector`. However the the exported CSS class will still be exported to Javascript.
+
+## Use cases
+
+### Container and recursive components
+
+Some components might contain nested instances of themselves, because they're a container or a "recursive" component (e.g. Tree component might render itself). 
+
+In case the component exposes any `pseudo-elements`, it is a good practice to define them as `custom-selectors` with [child selectors](https://developer.mozilla.org/en-US/docs/Web/CSS/Child_selectors) to avoid effecting internal instances inner parts.
+
+The following example show how a tree component exposes an icon:
+
+```css
+@namespace "Tree";
+@custom-selector :--icon .root > .icon;
+```
+
+Using the icon `pseudo-selector` from the outside just like a CSS class selector:
+```css
+@namespace "Panel";
+:import {
+    -st-from: "./tree.st.css";
+    -st-default: Tree;
+}
+/*
+selector: .Panel__root .Tree__root > .Tree__icon
+*/
+Tree::icon {
+    background: yellow;
+}
+```
+
+### Expose deep inner parts
+
+When you want to make some internal parts more accessible in your component API, you can simply describe `pseudo-elements` using `custom selectors`.
+
+For example exposing a `pseudo-element` named `navigationBtn` that allow to style an internal gallery component `navBtn` element:
+```css
+:import {
+    -st-from: "./gallery.st.css";
+    -st-default: Gallery;
+}
+@custom-selector :--navigationBtn Gallery::navBtn;
+```
+
+### Combination selector
+
+Sometimes a component has several basic CSS classes (with corresponding `pseudo-elements`) and we want to expose a combination `pseudo-element`.
+
+For example a `pseudo-element` named `navBtn` will match any `btn` CSS class nested in a `nav` CSS class:
+```css
+@namespace "Comp";
+@custom-selector :--navBtn .nav .btn;
+```
+
+```css
+@namespace "Page";
+:import {
+    -st-from: "./comp.st.css";
+    -st-default: Comp;
+}
+/*
+selector: Page__root .Comp__root .Comp__nav .Comp__btn
+*/
+Comp::navBtn { 
+    border: 1px solid grey; 
+}
+```
+
+### Selectors group
+
+In some cases you might want to gather up a collection of selectors into a single selector.
+
+```css
+@namespace "Comp";
+@custom-selector :--symbol .icon, .thumb, .picture;
+/*
+selector: 
+.Comp__root .Comp__icon, 
+.Comp__root .Comp__thumb, 
+.Comp__root .Comp__picture
+*/
+:--heading { 
+    border: 1px solid grey; 
+}
+```
+
+#### Caveats
+
+Aliasing multiple selectors in a `custom-selector` is a [footgun feature](https://en.wiktionary.org/wiki/footgun#English) that might generate lots of CSS.
```

Comment:
Should the user care about this "footgun feature"? Why should they care that lots of CSS is generated? File size for text is trivial

Reply:
Once you group multiple selectors under a custom selector alias you get the same duplication of selectors anywhere you use it. This is not that bad.

But because that custom selector is exposed as a custom pseudo element, a user styling it from the outside has no idea that everytime he target that pseudo element he's selectors are multiplied.

This gets worse for every level you go up...

Reply:
Yup, but my point is: does it matter? I believe that the performance of those selector is effectively free, and the resultant CSS, when minified and gzipped, is unlikely to bloat significantly. The cost of downloading a CSS file (compared with hero images, huge slabs of JS to parse then exectute) is pretty trivial (88K of the average page's 3.3 MB http://httparchive.org/interesting.php#bytesperpage). If it doesn't matter, don't mention it.

Reply:
The problem is that the output is unpredictably big from a user position.

Example with 3 stylesheet each composition the other. Each describes a custom selector that expand the previous one with 2 groups. And at the end an application stylesheet that imports the top stylesheet and styling the pseudo-element that was generated by the custom selector:

```css
@namespace "A";
@custom-selector :--x .a1, .a2;
```
```css
@namespace "B";
:import {
    -st-from: "a.st.css";
    -st-default: A;
}
@custom-selector :--y .b1 A::x, .b2 A::x;
```
```css
@namespace "C";
:import {
    -st-from: "b.st.css";
    -st-default: B;
}
@custom-selector :--z .c1 B::y, .c2 B::y;
:--z { color: red; }
```
```css
@namespace "App";
:import {
    -st-from: "c.st.css";
    -st-default: C;
}
C::z { color: red; } /* only ruleset */
```

OUTPUT:
```css
.App--root .C--root .C--c1 .B--root .B--b1 .A--root .A--a1 { color: red; }
.App--root .C--root .C--c1 .B--root .B--b1 .A--root .A--a2 { color: red; } 
.App--root .C--root .C--c1 .B--root .B--b2 .A--root .A--a1 { color: red; }
.App--root .C--root .C--c1 .B--root .B--b2 .A--root .A--a2 { color: red; }
.App--root .C--root .C--c2 .B--root .B--b1 .A--root .A--a1 { color: red; }
.App--root .C--root .C--c2 .B--root .B--b1 .A--root .A--a2 { color: red; } 
.App--root .C--root .C--c2 .B--root .B--b2 .A--root .A--a1 { color: red; }
.App--root .C--root .C--c2 .B--root .B--b2 .A--root .A--a2 { color: red; }
```

Reply:
Again, I'm not sure what the problem is. The resulting CSS is 8 lines; the Stylable source code is 10 lines. I think pointing out "footguns" should only be done in cases of extreme bloat/ performance problems. In this case, it's likely to deter people from using the library.

Reply:
lines in stylable source- but only 1 actual CSS delaration